### PR TITLE
Reintegrate viperserver backend

### PR DESCRIPTION
**Pull request by jogasser**
# Reintegrate viperserver backend
- readds viperserver as a dependency
- adds ViperServerWithSilicon(VSWITHSILICON) and ViperServerWithCarbon(VSWITHCARBON) as potential backends to choose from
- adds cacheFile config option to specify a cache file to be used by viperserver

### CODE REVIEW

File: src/main/scala/viper/gobra/frontend/Config.scala [hunk 1]
```diff
@@ -211,6 +214,13 @@ class ScallopGobraConfig(arguments: Seq[String], isInputOptional: Boolean = fals
     noshort = false
   )
 
+  val cacheFile: ScallopOption[String] = opt[String](
+    name = "cacheFile",
+    descr = "Cache file to be used, only usable if using viper server as a backend",
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Config.scala:219]: Would shorten to "Cache file to be used by Viper Server".

File: src/main/scala/viper/gobra/frontend/Config.scala [hunk 2]
```diff
@@ -295,6 +305,14 @@ class ScallopGobraConfig(arguments: Seq[String], isInputOptional: Boolean = fals
   }
   validateInput(input, module, include)
 
+  // cache file should only be usable when using viper server
+  validateOpt (backend, cacheFile) {
+    case (Some(ViperBackends.ViperServerWithSilicon), Some(_)) => Right()
+    case (Some(ViperBackends.ViperServerWithCarbon), Some(_)) => Right()
+    case (_, None) => Right()
+    case (_, Some(_)) => Left("Cache file option only works with viper server as a backend")
```
Comment (jcp19) [inline src/main/scala/viper/gobra/frontend/Config.scala:313]: Would change to "Cache file can only be specified when the backend uses Viper Server"

File: src/main/scala/viper/gobra/backend/ViperBackends.scala [hunk 3]
```diff
@@ -26,12 +31,75 @@ object ViperBackends {
...
+
+  object ViperServerWithSilicon extends ViperBackend {
+    def create(exePaths: Vector[String], config: Config)(implicit executor: GobraExecutionContext): ViperServer = {
+      if(ViperServerBackend.executor == null) {
+        ViperServerBackend.setExecutor(executor)
+      }
+
+      if(ViperServerBackend.server == null) {
+        ViperServerBackend.initServer(config)
+      }
```
Comment (jcp19) [inline src/main/scala/viper/gobra/backend/ViperBackends.scala:79]: I wonder if we should have these if statements surrounding the calls to `setExecutor` and `initServer`. This looks like a "silent failure" and might become a bit confusing for debugging later. Does it even make sense to not fail here?

File: src/main/scala/viper/gobra/backend/ViperBackends.scala [hunk 4]
```diff
@@ -26,12 +31,75 @@ object ViperBackends {
...
+
+  object ViperServerWithCarbon extends ViperBackend {
+    def create(exePaths: Vector[String], config: Config)(implicit executor: GobraExecutionContext): ViperServer = {
+      if(ViperServerBackend.executor == null) {
+        ViperServerBackend.setExecutor(executor)
+      }
+
+      if(ViperServerBackend.server == null) {
+        ViperServerBackend.initServer(config)
+      }
```
Comment (jcp19) [inline src/main/scala/viper/gobra/backend/ViperBackends.scala:99]: same as above

File: viper-toolchain-versions.sh [hunk 5]
```diff
@@ -8,3 +8,4 @@
 SILVER_REF="f645539a04e76d70337d7b5b108e7f9a9de9bc7a"
 SILICON_REF="v.21.07-release"
 CARBON_REF="v.21.07-release"
+VIPERSERVER_REF="master"
```
Comment (jcp19) [inline viper-toolchain-versions.sh:11]: Is there any reason not to use version "v.21.07-release" like with the other dependencies?

File: src/test/scala/viper/gobra/GobraTests.scala [hunk 6]
```diff
@@ -29,12 +30,17 @@ class GobraTests extends AbstractGobraTests with BeforeAndAfterAll {
...
 
   override def beforeAll(): Unit = {
+    executor = new DefaultGobraExecutionContext()
     gobraInstance = new Gobra()
   }
 
   override def afterAll(): Unit = {
+    executor.terminateAndAssertInexistanceOfTimeout()
+    ViperServerBackend.resetServer();
+    ViperServerBackend.resetExecutor();
```
Comment (jcp19) [inline src/test/scala/viper/gobra/GobraTests.scala:43]: why do we do this here?

**Review by jcp19 — CHANGES_REQUESTED** [review]
In general, looks good but I would add Viper Server as a dependency in our ReadMe, in the installation instructions

File: src/main/scala/viper/gobra/backend/ViperBackends.scala (hunk 3 above)
Comment (jogasser) [reply src/main/scala/viper/gobra/backend/ViperBackends.scala:79]: Yes, the goal here is to initialize the ViperServer environment if it wasn't previously. `create` gets called for each verification, if we do it like this, we can use a single ViperServer instance for multiple verification runs.

File: viper-toolchain-versions.sh (hunk 5 above)
Comment (jogasser) [reply viper-toolchain-versions.sh:11]: There will be after https://github.com/viperproject/viperserver/pull/44 is merged. This then needs to be changed to the new commit

File: src/main/scala/viper/gobra/backend/ViperBackends.scala (hunk 3 above)
Comment (jogasser) [reply src/main/scala/viper/gobra/backend/ViperBackends.scala:79]: I updated it to better handle executor shutdowns

File: src/test/scala/viper/gobra/GobraTests.scala (hunk 6 above)
Comment (jogasser) [reply src/test/scala/viper/gobra/GobraTests.scala]: I actually updated `ViperBackends` such that we don't need this here

File: src/main/scala/viper/gobra/backend/ViperBackends.scala (hunk 3 above)
Comment (jcp19) [reply src/main/scala/viper/gobra/backend/ViperBackends.scala]: > Yes, the goal here is to initialize the ViperServer environment if it wasn't previously. create gets called for each verification, if we do it like this, we can use a single ViperServer instance for multiple verification runs.

I would add a comment in the code explaining exactly what you said

**Comment by jcp19** [comment]
PS: we should adapt the [README](https://github.com/viperproject/gobra/blob/master/README.md) with the new dependency

File: src/test/scala/viper/gobra/GobraTests.scala [hunk 7]
```diff
@@ -9,6 +9,7 @@ package viper.gobra
 import java.nio.file.Path
 import ch.qos.logback.classic.Level
 import org.scalatest.BeforeAndAfterAll
+import viper.gobra.backend.ViperBackends.ViperServerBackend
```
Comment (jcp19) [inline src/test/scala/viper/gobra/GobraTests.scala:12]: This imports does not seem to be needed

File: src/main/scala/viper/gobra/backend/ViperServer.scala [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+import scala.concurrent.{Await, Future, Promise}
+import viper.gobra.util.GobraExecutionContext
+import viper.server.core.{CarbonConfig, SiliconConfig, VerificationExecutionContext, ViperBackendConfig, ViperCoreServer, ViperServerBackendNotFoundException}
+
+import scala.concurrent.duration.Duration
+
+object ViperServer {
+
+  case object Result
```
Comment (jcp19) [inline src/main/scala/viper/gobra/backend/ViperServer.scala:22]: what is the purpose of this `Result` object? Is this even used?

**Review by jcp19 — APPROVED** [review]
LGTM

**Comment by ArquintL** [comment]
@jogasser can you please update Silicon to commit [`22551b40509c381991a493da0108ea4c97fd602d`](https://github.com/viperproject/silicon/commit/22551b40509c381991a493da0108ea4c97fd602d)? Thanks!

**Review by jcp19 — COMMENTED** [review]
Looks fine to me

**Pull request merged**